File: NeuroflowLib/supervised_sample.cpp
#include "stdafx.h"
#include "supervised_sample.h"

USING;

supervised_sample::supervised_sample(const supervised_sample_entry& entry)
{
    add(entry);
}

supervised_sample::supervised_sample(const data_array_ptr& input)
{
    add(input);
}

supervised_sample::supervised_sample(const data_array_ptr& input, const data_array_ptr& desiredOutput, const data_array_ptr& actualOutput)
{
    add(input, desiredOutput, actualOutput);
}

idx_t supervised_sample::number_of_outputs() const
{
    idx_t c = 0;
    for (auto& e : _entries) if (e.has_output()) c++;
    return c;
}

std::list<supervised_sample_entry>& supervised_sample::entries()
{
    return _entries;
}

void supervised_sample::add(const supervised_sample_entry& entry)
{
    _entries.emplace_back(entry);
}

void supervised_sample::add(const data_array_ptr& input)
{
    _entries.emplace_back(input);
}

void supervised_sample::add(const data_array_ptr& input, const data_array_ptr& desiredOutput, const data_array_ptr& actualOutput)
{
    _entries.emplace_back(input, desiredOutput, actualOutput);
}

File: _vault/prev_version_win/Neuroflow/Data/SupervisedSample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Neuroflow.Data
{
    public sealed class SupervisedSample : List<SupervisedSampleEntry>, IDisposable
    {
        public SupervisedSample()
        {
        }

        public SupervisedSample(IEnumerable<SupervisedSampleEntry> collection) :
            base(collection)
        {
        }

        public SupervisedSample(SupervisedSampleEntry entry)
        {
            Args.Requires(() => entry, () => entry != null);

            Add(entry);
        }

        public SupervisedSample(DataArray input)
        {
            Add(input);
        }

        public SupervisedSample(DataArray input, DataArray desiredOutput, DataArray actualOutput)
        {
            Add(input, desiredOutput, actualOutput);
        }

        public int NumberOfOutputs
        {
            get { return this.Count(i => i.HasOutput); }
        }

        public void Add(DataArray input)
        {
            Add(new SupervisedSampleEntry(input));
        }

        public void Add(DataArray input, DataArray desiredOutput, DataArray actualOutput)
        {
            Add(new SupervisedSampleEntry(input, desiredOutput, actualOutput));
        }

        bool disposed;

        public void Dispose()
        {
            if (!disposed)
            {
                disposed = true;

                foreach (var da in this) da.Dispose();

                GC.SuppressFinalize(this);
            }
        }
    }
}
